graceful/service.go

```diff
@@ -117,36 +135,82 @@ func (s *Service) listenAndServe(ctx context.Context, _ string, addr string, ser
 		ln = tls.NewListener(ln, server.TLSConfig)
 	}
 
-	s.httpServer = server
-
 	go func() {
 		err := server.Serve(ln)
 		if !errors.Is(err, http.ErrServerClosed) {
 			log.WithError(err).Error("http server serve")
 		}
 	}()
 
+	if curServerCount+1 == s.numServers {
+		err := s.finalize(ctx)
+		if err != nil {
+			// purposefully do not wrap error here, as it is wrapped in finalize
+			return err
+		}
+	}
+
+	return nil
+}
+
+// prepare is called before the first server is started.
+func (s *Service) prepare(ctx context.Context) error {
+	if s.pidFile != "" {
+		err := os.Remove(s.pidFile)
+		if err != nil && !os.IsNotExist(err) {
+			return errors.Wrap(ctx, err, "fail to remove PID file")
```

Comment:
We don't begin error message by "fail to" any more
```suggestion
			return errors.Wrap(ctx, err, "remove PID file")
```